File: backend/services/ats_scorer.py
import re
from collections import Counter


ATS_SECTIONS = [
    "professional summary", "summary", "objective", "profile",
    "work experience", "experience", "employment",
    "education",
    "skills", "technical skills", "core competencies",
    "certifications", "projects", "awards", "publications",
]

PROBLEMATIC_PATTERNS = [
    r"\btable\b", r"\bcolumn\b", r"\btext box\b", r"\bgraphic\b",
    r"\bheader\b", r"\bfooter\b", r"\bimage\b",
]

STOP_WORDS = {
    "the", "a", "an", "and", "or", "but", "in", "on", "at", "to", "for",
    "of", "with", "by", "from", "is", "are", "was", "were", "be", "been",
    "being", "have", "has", "had", "do", "does", "did", "will", "would",
    "could", "should", "may", "might", "must", "shall", "this", "that",
    "these", "those", "i", "you", "he", "she", "we", "they", "it",
}


def compute_ats_score(resume_text: str, job_description: str) -> dict:
    """Compute a mock ATS score with detailed breakdown."""

    resume_lower = resume_text.lower()
    jd_lower = job_description.lower()

    # --- Keyword Match Score (40 pts) ---
    jd_tokens = set(re.findall(r"\b[a-zA-Z][a-zA-Z0-9+#.]{2,}\b", jd_lower)) - STOP_WORDS
    resume_tokens = set(re.findall(r"\b[a-zA-Z][a-zA-Z0-9+#.]{2,}\b", resume_lower)) - STOP_WORDS

    matched_keywords = jd_tokens & resume_tokens
    keyword_score = min(40, int((len(matched_keywords) / max(len(jd_tokens), 1)) * 80))

    top_jd_keywords = [w for w in jd_tokens if len(w) > 4][:20]
    matched_top = [w for w in top_jd_keywords if w in resume_tokens]
    missing_keywords = [w for w in top_jd_keywords if w not in resume_tokens][:10]

    # --- Section Detection Score (25 pts) ---
    detected_sections = []
    for section in ATS_SECTIONS:
        if section in resume_lower:
            detected_sections.append(section)

    section_score = min(25, int((len(detected_sections) / 5) * 25))

    # --- Formatting Score (20 pts) ---
    formatting_score = 20
    formatting_issues = []
    for pattern in PROBLEMATIC_PATTERNS:
        if re.search(pattern, resume_lower):
            formatting_score -= 3
            formatting_issues.append(f"Possible problematic element detected: '{pattern}'")

    # Check for bullet points (good)
    bullet_count = len(re.findall(r"^[•\-\*▪–]", resume_text, re.MULTILINE))
    if bullet_count < 3:
        formatting_score -= 5
        formatting_issues.append("Very few bullet points detected — ATS prefers structured bullet points")

    formatting_score = max(0, formatting_score)

    # --- Length Score (15 pts) ---
    word_count = len(resume_text.split())
    if 300 <= word_count <= 800:
        length_score = 15
        length_note = f"Good length ({word_count} words)"
    elif word_count < 300:
        length_score = 8
        length_note = f"Resume may be too short ({word_count} words) — aim for 300–800 words"
    else:
        length_score = 10
        length_note = f"Resume may be too long ({word_count} words) — consider trimming"

    total_score = keyword_score + section_score + formatting_score + length_score

    return {
        "total_score": total_score,
        "max_score": 100,
        "grade": _grade(total_score),
        "breakdown": {
            "keyword_match": {"score": keyword_score, "max": 40, "matched_count": len(matched_top), "total_jd_keywords": len(top_jd_keywords)},
            "section_structure": {"score": section_score, "max": 25, "detected_sections": detected_sections},
            "formatting": {"score": formatting_score, "max": 20, "issues": formatting_issues},
            "length": {"score": length_score, "max": 15, "note": length_note, "word_count": word_count},
        },
        "matched_keywords": matched_top,
        "missing_keywords": missing_keywords,
    }


def _grade(score: int) -> str:
    if score >= 85:
        return "A"
    elif score >= 70:
        return "B"
    elif score >= 55:
        return "C"
    elif score >= 40:
        return "D"
    else:
        return "F"

File: frontend/src/lib/ats-scorer.js
// Browser-side ATS scoring — ported from Python backend
const STOP_WORDS = new Set([
  'the','a','an','and','or','but','in','on','at','to','for','of','with',
  'by','from','is','are','was','were','be','been','being','have','has',
  'had','do','does','did','will','would','could','should','may','might',
  'must','shall','this','that','these','those','i','you','he','she','we',
  'they','it','as','up','about','into','through','during','including',
  'until','against','among','throughout','despite','towards','upon',
]);

const ATS_SECTIONS = [
  'professional summary','summary','objective','profile',
  'work experience','experience','employment',
  'education',
  'skills','technical skills','core competencies',
  'certifications','projects','awards','publications',
];

const PROBLEMATIC_PATTERNS = [/\btable\b/i, /\bcolumn\b/i, /\btext box\b/i, /\bgraphic\b/i];

function tokenize(text) {
  return new Set(
    (text.toLowerCase().match(/\b[a-zA-Z][a-zA-Z0-9+#.]{2,}\b/g) || [])
      .filter(w => !STOP_WORDS.has(w))
  );
}

function gradeScore(score) {
  if (score >= 85) return 'A';
  if (score >= 70) return 'B';
  if (score >= 55) return 'C';
  if (score >= 40) return 'D';
  return 'F';
}

export function computeAtsScore(resumeText, jobDescription) {
  const resumeLower = resumeText.toLowerCase();
  const resumeTokens = tokenize(resumeText);
  const jdTokens = tokenize(jobDescription);

  // ── Keyword Match (40 pts) ──────────────────────
  const topJdKeywords = [...jdTokens].filter(w => w.length > 4).slice(0, 20);
  const matchedTop = topJdKeywords.filter(w => resumeTokens.has(w));
  const missingKeywords = topJdKeywords.filter(w => !resumeTokens.has(w)).slice(0, 10);
  const keywordScore = Math.min(40, Math.round((matchedTop.length / Math.max(topJdKeywords.length, 1)) * 80));

  // ── Section Structure (25 pts) ──────────────────
  const detectedSections = ATS_SECTIONS.filter(s => resumeLower.includes(s));
  const sectionScore = Math.min(25, Math.round((detectedSections.length / 5) * 25));

  // ── Formatting (20 pts) ─────────────────────────
  let formattingScore = 20;
  const formattingIssues = [];
  PROBLEMATIC_PATTERNS.forEach(p => {
    if (p.test(resumeLower)) {
      formattingScore -= 3;
      formattingIssues.push(`Possible problematic element: ${p.source}`);
    }
  });
  const bulletCount = (resumeText.match(/^[•\-\*▪–]/gm) || []).length;
  if (bulletCount < 3) {
    formattingScore -= 5;
    formattingIssues.push('Very few bullet points — ATS prefers structured bullet points');
  }
  formattingScore = Math.max(0, formattingScore);

  // ── Length (15 pts) ─────────────────────────────
  const wordCount = resumeText.split(/\s+/).filter(Boolean).length;
  let lengthScore, lengthNote;
  if (wordCount >= 300 && wordCount <= 800) {
    lengthScore = 15;
    lengthNote = `Good length (${wordCount} words)`;
  } else if (wordCount < 300) {
    lengthScore = 8;
    lengthNote = `Resume may be too short (${wordCount} words) — aim for 300–800 words`;
  } else {
    lengthScore = 10;
    lengthNote = `Resume may be too long (${wordCount} words) — consider trimming`;
  }

  const totalScore = keywordScore + sectionScore + formattingScore + lengthScore;

  return {
    total_score: totalScore,
    max_score: 100,
    grade: gradeScore(totalScore),
    breakdown: {
      keyword_match: { score: keywordScore, max: 40, matched_count: matchedTop.length, total_jd_keywords: topJdKeywords.length },
      section_structure: { score: sectionScore, max: 25, detected_sections: detectedSections },
      formatting: { score: formattingScore, max: 20, issues: formattingIssues },
      length: { score: lengthScore, max: 15, note: lengthNote, word_count: wordCount },
    },
    matched_keywords: matchedTop,
    missing_keywords: missingKeywords,
  };
}
